components/appmgt/org.wso2.carbon.appmgt.services.api/src/main/java/org/wso2/carbon/appmgt/services/api/v1/apps/discovery/service/DiscoveryService.java

```diff
@@ -0,0 +1,202 @@
+
+    /**
+     * Returns the full application information so that an application can be created with the
+     * information.
+     *
+     * @return
+     */
+    @POST
+    @Path("app/info/{tenantDomain}")
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces(MediaType.APPLICATION_JSON)
+    public DiscoveredApplicationDTO getDiscoveredApplication(
+            @Context final HttpServletRequest servletRequest,
+            @Context final HttpServletResponse servletResponse, @Context HttpHeaders headers,
+            @PathParam("tenantDomain") String tenantDomain, DiscoveryInfoRequest discoverInfoRequest) {
+
+        if (tenantDomain == null)
+            tenantDomain = SUPER_TENANT_DOMAIN;
+        DiscoveredApplicationDTO result = null;
+
+        ApplicationDiscoveryServiceFactory applicationDiscoveryServiceFactory = (ApplicationDiscoveryServiceFactory) PrivilegedCarbonContext
+                .getThreadLocalCarbonContext()
+                .getOSGiService(ApplicationDiscoveryServiceFactory.class);
+        ApplicationDiscoveryHandler handler = applicationDiscoveryServiceFactory
+                .getHandler("WSO2-AS");
+
+        APIIdentifier apiIdentifier = new APIIdentifier(null, null, null);
+        apiIdentifier.setApplicationId(discoverInfoRequest.getApplicationId());
+
+        PrivilegedCarbonContext carbonContext = PrivilegedCarbonContext
+                .getThreadLocalCarbonContext();
+        ApplicationDiscoveryContext applicationDiscoveryContext = (ApplicationDiscoveryContext) servletRequest.getSession().getAttribute(ApplicationDiscoveryContext.class.getName());
+        if(applicationDiscoveryContext == null) {
+            applicationDiscoveryContext = new ApplicationDiscoveryContext();
+            servletRequest.getSession().setAttribute(ApplicationDiscoveryContext.class.getName(), applicationDiscoveryContext);
+        }
+        try {
+            result = handler
+                    .readApplicationInfo(applicationDiscoveryContext, apiIdentifier,
+                             carbonContext);
+        } catch (AppManagementException e) {
+            String message = "Error while discovering the application from the backend server Server[%s], User[%s], Reason: ";
+            log.error(message);
+            log.debug(message, e);
+            safeSendError(servletResponse, Response.Status.INTERNAL_SERVER_ERROR);
+        }
+
+        return result;
+    }
+
+    /**
+     * Returns the list of applications, number of pages.
+     * This will return only the minimal required fields for listing to conserve memory and bandwidth.
+     *
+     * sample request
+     *  {"credentials" : {"userName" : "admin", "appServerUrl" : "local", "password" : "admin" ,
+     *  "loggedInUsername" : "admin"},
+     *  "searchCriteria" : {"applicationName" : "", "status" : "New", "pageNumber" : 4 }
+     *  }
+     *
+     * @return
+     */
+    @POST
+    @Path("app/list/{tenantDomain}")
```

Comment:
This API seems not following RESTful best practices. Better to redesign the API.